### ISSUE TITLE
Calling Method Within Struct Impl

### ISSUE THREAD

**Issue by esheri3** [drrb/java-rust-example#2]
Thank you for taking the time to put together this example project. This has been very useful for my introduction into calling Rust libs from Java. Can you please provide an example illustrating the changes required to allow a Java program to call a function declared within a Struct impl? For example, assume you add a new fn called "string" within the existing "Greeting" impl that accepts a param called length (u32 or whatever) and returns a const char * that acts as a String in Java. How would I call this on the Java side? Would I create a new "string" method in Greeting.java that matches the signature? Is this even possible? I'm trying to understand if I can preserve the OOP abstractions from Rust in Java or if I have to have all C-like flat function representations for Java/JNA to work.

**Comment by drrb** [comment]
No worries! I'm glad it's been useful. I'll try to take a look at creating an example like that.

I don't think it's possible to directly map a method on a struct to the equivalent method in Rust, though I could definitely be wrong. What I've opted for in my own code is creating a plain Java object to wrap the JNA Structure object. The outer object contains an instance of the JNA struct, and it passes it to the functions on the Library interface internally. Then, in Rust, the entry-point functions call the method on the Rust struct that was passed from Java. So you still have the C-like functions on the Library interface, but your code can interact with the Java wrapper object almost like it was interacting with the Rust struct. I've hastily typed up an example [gist](https://gist.github.com/drrb/9e6ad391c58e0aa00821) (probably doesn't work as is, I typed it straight into the browser :) ).
